[JSC] Add VectorTraits for WriteBarrier

Constellation opened the pull request:
#### 1eb3c9f9ce70458c2df143468ece882e041cdd6f
<pre>
[JSC] Add VectorTraits for WriteBarrier
<a href="https://bugs.webkit.org/show_bug.cgi?id=196242">https://bugs.webkit.org/show_bug.cgi?id=196242</a>

Reviewed by Darin Adler.

WriteBarrier&lt;&gt; is one of the most frequently used class for Vector etc. in JSC.
Let&apos;s add VectorTraits for that.

* Source/JavaScriptCore/runtime/WriteBarrier.h:

Canonical link: <a href="https://commits.webkit.org/252556@main">https://commits.webkit.org/252556@main</a>
</pre>

darinadler commented on Source/JavaScriptCore/runtime/WriteBarrier.h:
```diff
@@ -352,3 +352,24 @@ class WriteBarrierStructureID {
+
+template<typename T> struct VectorTraits<JSC::WriteBarrier<T>> : public SimpleClassVectorTraits {
+    static_assert(std::is_trivially_destructible<JSC::WriteBarrier<T>>::value);
+    static const bool canInitializeWithMemset = true;
```
Since `canInitializeWithMemset` is already `true` in the base class, I think we should omit this line.

Constellation replied:
Right. Fixed.

darinadler commented on Source/JavaScriptCore/runtime/WriteBarrier.h:
```diff
@@ -352,3 +352,24 @@ class WriteBarrierStructureID {
+template<> struct VectorTraits<JSC::WriteBarrier<JSC::Unknown>> : public SimpleClassVectorTraits {
+    static_assert(std::is_trivially_destructible<JSC::WriteBarrier<JSC::Unknown>>::value);
+#if USE(JSVALUE64)
+    // We can memset only in JSVALUE64 since empty value is zero. On the other hand, JSVALUE32_64's empty value is not zero.
```
Comment is great but code should be `#if !USE(JSVALUE64)` and only include setting to `false`, rely on inheritance for the true case.

Constellation replied:
Right, fixed.

darinadler commented on Source/JavaScriptCore/runtime/WriteBarrier.h:
```diff
@@ -352,3 +352,24 @@ class WriteBarrierStructureID {
+    static const bool canInitializeWithMemset = false;
+#endif
+    static const bool canCopyWithMemcpy = true;
+};
```
Can we also set `canCompareWithMemcmp` to `true`? Also seems like `operator==` for two `WriteBarrier<Unknown>` should not decode the two values, just compare them still encoded.

Constellation replied:
Rigt, changed :)

Constellation replied:
Ah, `canCompareWithMemcmp` was true since `SimpleClassVectorTraits` is setting it to true :)

Constellation committed:
[JSC] Add VectorTraits for WriteBarrier
https://bugs.webkit.org/show_bug.cgi?id=196242

Reviewed by Darin Adler.

WriteBarrier<> is one of the most frequently used class for Vector etc. in JSC.
Let's add VectorTraits for that.

* Source/JavaScriptCore/runtime/WriteBarrier.h:

Canonical link: https://commits.webkit.org/252556@main

justinmichaud commented on Source/JavaScriptCore/runtime/WriteBarrier.h:
```diff
@@ -352,3 +352,21 @@ class WriteBarrierStructureID {
+
+template<typename T> struct VectorTraits<JSC::WriteBarrier<T>> : public SimpleClassVectorTraits {
+    static_assert(std::is_trivially_destructible<JSC::WriteBarrier<T>>::value);
+    static constexpr bool canCopyWithMemcpy = true;
```
@Constellation Isn't this wrong? My understanding is that memcpy is free to use load/store instructions with fewer than 8 bytes, which would cause GC tearing. I know we do something special for some copies, what is the current state of that?
